onenetd.1
.TH onenetd 1 "2026-01-31" "Version 1.11c" "Network Tools"
.SH NAME
onenetd \- listen on a TCP port and launch a program to handle requests
.SH SYNOPSIS
onenetd [options] ADDRESS PORT COMMAND ...
.SH DESCRIPTION
onenetd binds a TCP socket to port PORT on IP address ADDRESS, listens for and
accepts connections, and spawns COMMAND with stdin and stdout (and optionally
stderr) connected to the accepted socket.
.PP
onenetd works like a minimal version of inetd that listens for
 "stream tcp nowait" services on a single port. If you need
to listen on mutiple ports, just run multiple copies of onenetd, 
one for each service.
.PP
The advantages of onenetd over inetd include:
.IP \(bu
Connections may be refused with a fixed response after the connection limit is
reached.
.IP \(bu
TCP_NODELAY may be set on a per-port basis.
.IP \(bu
The address to bind to may be chosen arbitrarily, so it's easy to make a
service only listen on localhost, or run several different services on the same
port for different interface addresses.
.IP \(bu
Incoming connections may be logged to stderr.
.PP
Note that onenetd only handles TCP connections, so you can't run UDP
services such as nmbd.
.SH OPTIONS
.TP
.B ADDRESS
The hostname or IP address to bind() to. Specify 0 to bind to all addresses.
.TP
.B PORT
The service name or port number to bind() to. If you need to run onenetd as root
to bind to a privileged port, you can use the options -u/-g or -U to switch uid
and gid after bind()ing.
.TP
.B COMMAND ...
The command to execute, as well as its parameters (all parameters given to
onenetd after COMMAND are passed to COMMAND as parameters). argv[0] for COMMAND
will be set to COMMAND; if you need to specify a different argv[0], use the
argv0 tool from the freedt or daemontools packages.
.TP
.B \-c N
Limit the number of connections (and thus child processes) to N. The default is
40.
.TP
.B \-n num
Set the starting value of the CONNCOUNT environment variable to num. The
default is 0.
.TP
.B \-g GID
After binding the TCP socket, setgid(GID). GID must be numeric.
.TP
.B \-u UID
After binding the TCP socket, setuid(UID). UID must be numeric.
.TP
.B \-U
After binding the TCP socket, setgid($GID) and setuid($UID). $GID and $UID must
be numeric. This is intended for use with envuidgid (from daemontools or
freedt).
.TP
.B \-b N
Set the listen() backlog to N. Note that on many operating systems, the listen
backlog has an arbitrary limit, or may be entirely ignored. The default is 10.
.TP
.B \-D
Set the TCP_NODELAY option on the accepted sockets. This causes the operating
system's TCP stack to avoid coalescing smaller packets into larger ones,
decreasing latency but reducing throughput.
.TP
.B \-Q
Do not print any messages to stderr, except in the case of a fatal error. This
is the default.
.TP
.B \-v
Print messages to stderr indicating clients connecting and disconnecting.
These can be collected for logging purposes using logger, multilog from
daemontools, or dumblog from freedt.
.TP
.B \-h
Print a brief usage message, then exit. The same message will be shown if
unknown or invalid arguments are passed.
.TP
.B \-r MESSAGE
Normally once the number of active connections has passed the limit set by -c,
further connections will not be accepted until the number of active connections
has dropped again -- that is, further connections will not be refused, but will
not connect until free connections are available. If -r is specified,
connections will be accepted and have the constant string MESSAGE printed to
them, then will be disconnected; this is done inside the onenetd process
without forking, so it does not use up process slots. MESSAGE may contain \\r,
\\n or \\t to specify a carriage return, newline or tab character.
.SH EXIT CODE
onenetd returns 0 on success, or 20 on failure (for instance, invalid
arguments). Since onenetd is intended to run forever, the exit code is
generally not important, and stderr should be examined for diagnostic messages
if it exits unexpectedly.
.SH ENVIRONMENT VARIABLES
onenetd sets several variables in the child process's environment. These can be
used for logging or access control.
.TP
.B TCPLOCALIP
The local IPv4 address of the connected socket, in dotted quad format.
.TP
.B TCPLOCALPORT
The local IPv4 port of the connected socket, as a decimal number.
.TP
.B TCPREMOTEIP
The remote IPv4 address of the connected socket, in dotted quad format.
.TP
.B TCPREMOTEPORT
The remote IPv4 port of the connected socket, as a decimal number.
.TP
.B CONNCOUNT
A unique sequence number for each connection. Can be used as a unique
identifier in filenames created by the spawned process, or in logfiles.
.SH EXAMPLES
.TP
onenetd 0 echo cat
Simulate the standard inetd echo service.
.TP
onenetd -v 192.168.1.2 daytime date
Simulate the standard inetd daytime service on the interface 192.168.1.2,
logging connections to stderr.
.TP
onenetd -v -c 5 -r '421 Server full\\r\\n' 0 21 /usr/sbin/ftpd -a /home/ftp
Run an FTP service with a limit of five concurrent connections, refusing any
further connections with an appropriate FTP error message. ftpd will be invoked
as "/usr/sbin/ftpd -a /home/ftp".
.TP
envuidgid myuser onenetd -v -U 0 25098 myprog --my-arguments
Set user and group to that of myuser, and run "myprog --my-arguments" for
connections to port 25098 on any interface.
.SH AUTHOR
onenetd was written by Adam Sampson <ats@offog.org>
.SH SEE ALSO
.BR inetd (8),
.BR argv0 (8),
.BR envuidgid (8),
.BR softlimit (8),
.BR dumblog (8),
.BR logger (1),
.BR sock (1).


NEWS
(This only includes features; for bugfixes, see ChangeLog.)

1.11c: Set CONNCOUNT to feed sequential connection numbers to scripts.
1.11: Set TCPLOCALIP to the real local address; document environment
      variables; change licence.
1.10: No functional changes.
1.9: No functional changes.
1.8: Add -e and manual page.
1.7: Fix behaviour after server has been full.
1.6: Always try to send response first; include child counts in status
     messages.
1.5: Remove test for host being "0"; made help better.
1.4: Fix linking under Solaris.
1.3: Removed -K.
1.2: Added escape characters in response.
1.1: Initial release.



onenetd.c
/*
   onenetd: a single-process inetd equivalent
   Copyright 2001, 2002, 2003, 2005 Adam Sampson <ats@offog.org>

   Tweaked from 1.11 to 1.11c, where 'c' stands for CONNCOUNT.
   See 'CONNCOUNT' in the code and the man page to explain.
   Paul Ducklin <pducklin@outlook.com>

   Permission is hereby granted, free of charge, to any person obtaining a
   copy of this software and associated documentation files (the
   "Software"), to deal in the Software without restriction, including
   without limitation the rights to use, copy, modify, merge, publish,
   distribute, sublicense, and/or sell copies of the Software, and to
   permit persons to whom the Software is furnished to do so, subject to
   the following conditions:

   The above copyright notice and this permission notice shall be included
   in all copies or substantial portions of the Software.

   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
   OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
   MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
   IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
   CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
   TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
   SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <netdb.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netinet/tcp.h>
#include <arpa/inet.h>
#include <sys/time.h>
#include <sys/types.h>
#include <signal.h>
#include <errno.h>
#include <fcntl.h>
#include <sys/wait.h>
#include <stdio.h>

#define VERSION "1.11c"

int max_conns = 40;
int conn_count = 0;
int conn_total = 0;
long gid = -1;
long uid = -1;
int backlog = 10;
int no_delay = 0;
int verbose = 0;
int stderr_to_socket = 0;
char *response = NULL;
char **command;
int selfpipe[2];
int listen_fd;

typedef struct client {
	int fd;
	char *message;
	int left;
	struct client *next;
} client;
client *clients = NULL;

/* Print a warning. */
void warn(const char *msg) {
	fprintf(stderr, "%s\n", msg);
}

/* Die with an error message. */
void die(const char *msg) {
	warn(msg);
	exit(20);
}

/* Handle SIGCHLD. */
void handle_sigchld(int dummy) {
	int old_errno = errno;
	write(selfpipe[1], "c", 1);
	errno = old_errno;
}

/* Change the flags on an fd. */
int change_flags(int fd, int add, int remove) {
	int flags = fcntl(fd, F_GETFL);
	int newflags;
	if (flags == -1)
		return -1;

	newflags = (flags | add) & ~remove;

	if (newflags != flags) {
		if (fcntl(fd, F_SETFL, flags) < 0)
			return -1;
	}
	
	return 0;
}

/* Set the FD_CLOEXEC flag on an fd. */
void set_fd_cloexec(int fd) {
	if (fcntl(fd, F_SETFD, FD_CLOEXEC) < 0)
		die("unable to set FD_CLOEXEC");
}

/* Add an fd to an FD_SET, updating a maximum. */
void fd_set_add(int fd, fd_set *fds, int *max) {
	FD_SET(fd, fds);
	if (fd > *max) *max = fd;
}

/* Print the usage message. */
void usage(int code) {
	fprintf(stderr, "onenetd version " VERSION "\n"
		"Copyright 2001, 2002, 2003, 2005 Adam Sampson <ats@offog.org>\n"
		"This is free software with ABSOLUTELY NO WARRANTY.\n\n"
		"Usage: onenetd [options] address port command ...\n"
		"  address  Address to bind to (specify 0 for any address)\n"
		"  port     TCP port to bind to\n"
		"  command  Command to execute\n"
		"Options:\n"
		"  -c N     limit to at most N children running (default 40).\n"
		"           Further connections will be deferred unless -r\n"
		"           is specified.\n"
		"  -n num   set starting CONNCOUNT to num (default 0).\n"
		"  -g gid   setgid(gid) after binding\n"
		"  -u uid   setuid(uid) after binding\n"
		"  -U       setuid($UID) and setgid($GID) after binding\n"
		"  -b N     set listen() backlog to N\n"
		"  -D       set TCP_NODELAY option on sockets\n"
		"  -e       redirect stderr of children to socket\n"
		"  -v       be verbose\n"
		"  -r resp  once -c limit is reached, refuse clients\n"
		"           with 'resp' rather than deferring them.\n"
		"           resp may contain \\r, \\n, \\t.\n"
		"  -h       show this usage message\n");
	exit(code);
}

/* Try to send a chunk of the response to a client. Remove the client
   from the list if we've sent all of it. */
void try_to_send(client *prev_cl, client *cl) {
	int remove = 0;
	int count = write(cl->fd, cl->message, cl->left);

	if (count >= 0) {
		cl->message += count;
		cl->left -= count;

		if (cl->left == 0)
			remove = 1;
	} else if (errno == EAGAIN) {
		/* ignorable error */
	} else {
		/* another error while writing */
		remove = 1;
	}

	if (remove) {
		close(cl->fd);
		if (prev_cl) {
			prev_cl->next = cl->next;
		} else {
			clients = cl->next;
		}
		free(cl);
	}
}

int main(int argc, char **argv) {
	struct sigaction sa;
	sigset_t sig_chld;
	struct sockaddr_in listen_addr;
	char *s, *r;
	int n;
	
	while (1) {
		char c = getopt(argc, argv, "+c:n:g:u:Ub:ODQvehr:");
		if (c == -1)
			break;
		switch (c) {
		case 'c':
			max_conns = atoi(optarg);
			break;
		case 'n':
			conn_total = atoi(optarg);
			break;
		case 'g':
			gid = atoi(optarg);
			break;
		case 'u':
			uid = atoi(optarg);
			break;
		case 'U':
			s = getenv("GID");
			if (!s)
				die("-U specified but no $GID");
			gid = atoi(s);
			s = getenv("UID");
			if (!s)
				die("-U specified but no $UID");
			uid = atoi(s);
			break;
		case 'b':
			backlog = atoi(optarg);
			break;
		case 'D':
			no_delay = 1;
			break;
		case 'Q':
			verbose = 0;
			break;
		case 'v':
			verbose = 1;
			break;
		case 'e':
			stderr_to_socket = 1;
			break;
		case 'h':
			usage(0);
			break;
		case 'r':
			r = response = malloc(strlen(optarg) + 1);
			if (!r)
				die("out of memory");
			for (s = optarg; *s != '\0'; s++) {
				if (*s == '\\') {
					s++;
					if (*s == 'r')
						*r++ = '\r';
					else if (*s == 'n')
						*r++ = '\n';
					else if (*s == 't')
						*r++ = '\t';
					else
						usage(20);
				} else {
					*r++ = *s;
				}
			}
			*r = '\0';
			break;
		default:
			usage(20);
		}
	}

	if ((argc - optind) < 3)
		usage(20);

	listen_addr.sin_family = AF_INET;

	s = argv[optind++];
	listen_addr.sin_addr.s_addr = inet_addr(s);
	if (listen_addr.sin_addr.s_addr == -1) {
		struct hostent *he = gethostbyname(s);
		if ((!he) || (he->h_addrtype != AF_INET)
			|| (he->h_addr == 0))
			die("unable to resolve listen host");
		listen_addr.sin_addr = *(struct in_addr *)he->h_addr;
	}

	s = argv[optind++];
	n = atoi(s);
	if (n == 0 && strcmp(s, "0") != 0) {
		struct servent *se = getservbyname(s, "tcp");
		if (!se)
			die("unable to resolve listen port");
		n = se->s_port;
	} else {
		n = htons(n);
	}
	listen_addr.sin_port = n;
	command = &argv[optind];

	if (pipe(selfpipe) < 0)
		die("unable to create self-pipe");
	if (change_flags(selfpipe[1], O_NONBLOCK, 0) < 0)
		die("unable to set O_NONBLOCK");
	set_fd_cloexec(selfpipe[0]);
	set_fd_cloexec(selfpipe[1]);

	listen_fd = socket(PF_INET, SOCK_STREAM, IPPROTO_TCP);
	if (listen_fd < 0)
		die("unable to create socket");
	if (change_flags(listen_fd, O_NONBLOCK, 0) < 0)
		die("unable to set O_NONBLOCK");
	set_fd_cloexec(listen_fd);
	n = 1;
	if (setsockopt(listen_fd, SOL_SOCKET, SO_REUSEADDR, &n, sizeof n) < 0)
		die("unable to set SO_REUSEADDR");
	if (bind(listen_fd, (struct sockaddr *)&listen_addr,
		sizeof listen_addr) < 0)
		die("unable to bind to listen address");
	if (listen(listen_fd, backlog) < 0)
		die("unable to listen");

	if (gid != -1)
		if (setgid(gid) < 0)
			die("unable to setgid");
	if (uid != -1)
		if (setuid(uid) < 0)
			die("unable to setuid");

	sigemptyset(&sig_chld);
	sigaddset(&sig_chld, SIGCHLD);
	if (sigprocmask(SIG_BLOCK, &sig_chld, NULL) < 0)
		die("unable to block SIGCHLD");

	sa.sa_handler = handle_sigchld;
	sigemptyset(&sa.sa_mask);
	sa.sa_flags = SA_NOCLDSTOP;
	sigaction(SIGCHLD, &sa, NULL);

	while (1) {
		int full;
		client *cl, *prev_cl, *next_cl;
		fd_set read_fds, write_fds;

		do {
			sigset_t old_sigs;
			int max = -1;

			full = conn_count >= max_conns;
			FD_ZERO(&read_fds);
			fd_set_add(selfpipe[0], &read_fds, &max);
			if (!(full && !response))
				fd_set_add(listen_fd, &read_fds, &max);

			FD_ZERO(&write_fds);
			for (cl = clients; cl; cl = cl->next)
				fd_set_add(cl->fd, &write_fds, &max);

			if (sigprocmask(SIG_UNBLOCK, &sig_chld, &old_sigs) < 0)
				die("unable to unblock SIGCHLD");

			n = select(max + 1, &read_fds, &write_fds, NULL, NULL);
			if (n < 0 && errno != EINTR)
				warn("select failed");

			if (sigprocmask(SIG_SETMASK, &old_sigs, NULL) < 0)
				die("unable to restore signal mask");
		} while (n < 0);

		prev_cl = NULL;
		for (cl = clients; cl; cl = next_cl) {
			next_cl = cl->next;

			if (FD_ISSET(cl->fd, &write_fds))
				try_to_send(prev_cl, cl);
			prev_cl = cl;
		}

		if (FD_ISSET(selfpipe[0], &read_fds)) {
			char c;

			/* We don't care if this fails. */
			read(selfpipe[0], &c, 1);

			while (1) {
				n = waitpid(-1, NULL, WNOHANG);
				if (n <= 0) break;

				conn_count--;
				if (verbose) {
					fprintf(stderr, "        (%2d/%2d) "
						"pid %d closed\n",
						conn_count,max_conns,n);
				}
			}
		}

		if (FD_ISSET(listen_fd, &read_fds)) {
			int pid;
			struct sockaddr_in local_addr, child_addr;
			socklen_t len = sizeof child_addr;
			int child_fd;

			if (full && !response)
				goto no_conn;

			child_fd = accept(listen_fd,
				(struct sockaddr *)&child_addr, &len);
			
			if (len != sizeof child_addr) {
				warn("unable to get remote address");
				goto no_conn;
			}
			if (child_fd < 0 && errno == EAGAIN)
				goto no_conn;
			if (child_fd < 0) {
				warn("accept failed");
				goto no_conn;
			}
			set_fd_cloexec(child_fd);

			len = sizeof local_addr;
			if (getsockname(child_fd,
				(struct sockaddr *)&local_addr, &len) < 0
				|| len != sizeof local_addr) {
				warn("unable to get local address");
				goto no_conn;
			}

			if (full) {
				/* Avoid overfilling the fd_set. */
				if (child_fd >= FD_SETSIZE && verbose) {
					fprintf(stderr, "- dropped from %s "
						"port %d\n",
						inet_ntoa(child_addr.sin_addr),
						ntohs(child_addr.sin_port));
				}
				if (child_fd >= FD_SETSIZE)
					goto no_conn;

				if (change_flags(child_fd, O_NONBLOCK, 0) < 0) {
					warn("unable to set O_NONBLOCK");
					goto no_conn;
				}

				cl = malloc(sizeof *cl);
				if (!cl) {
					warn("out of memory");
					goto no_conn;
				}

				cl->fd = child_fd;
				child_fd = -1;
				cl->message = response;
				cl->left = strlen(cl->message);
				cl->next = clients;
				clients = cl;

				if (verbose)
					fprintf(stderr, "- refused from %s "
						"port %d\n",
						inet_ntoa(child_addr.sin_addr),
						ntohs(child_addr.sin_port));
	
				/* Try to send the response now; if we send
				   all of it it'll get removed from the list
				   again. */
				try_to_send(NULL, cl);

				goto no_conn;
			}

			n = 1;
			if (no_delay && setsockopt(child_fd, IPPROTO_TCP,
				TCP_NODELAY, &n, sizeof n) < 0) {
				warn("unable to set TCP_NODELAY");
				goto no_conn;
			}

			pid = fork();
			if (pid < 0) {
				warn("fork failed");
				goto no_conn;
			}

			if (pid == 0) {
				char buf[80];

				dup2(child_fd, 0);
				dup2(child_fd, 1);
				if (stderr_to_socket)
					dup2(child_fd, 2);

				putenv(strdup("PROTO=TCP"));
				snprintf(buf, sizeof buf, "CONNCOUNT=%d",
					conn_total);
				putenv(strdup(buf));
				snprintf(buf, sizeof buf, "TCPLOCALIP=%s",
					inet_ntoa(local_addr.sin_addr));
				putenv(strdup(buf));
				snprintf(buf, sizeof buf, "TCPLOCALPORT=%d",
					ntohs(local_addr.sin_port));
				putenv(strdup(buf));
				snprintf(buf, sizeof buf, "TCPREMOTEIP=%s",
					inet_ntoa(child_addr.sin_addr));
				putenv(strdup(buf));
				snprintf(buf, sizeof buf, "TCPREMOTEPORT=%d",
					ntohs(child_addr.sin_port));
				putenv(strdup(buf));

				execvp(command[0], command);
				_exit(20);
			}

			conn_count++;
			if (verbose) {
				fprintf(stderr, "%06d: (%2d/%2d) "
					"pid %d from %s:%d\n",
					conn_total,conn_count, max_conns,
					pid, inet_ntoa(child_addr.sin_addr),
					ntohs(child_addr.sin_port) );
			}
			conn_total++;

			no_conn:
			if (child_fd >= 0)
				close(child_fd);
		}	
	}

	return 0;
}

